### ISSUE TITLE
No error is reported when a lambda parameter is shadowing a range variable 

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#12053]
```
        var res1 = from a in new[] { 1 }
                   from x2 in new[] { 0 }
                   join b in new[] { 1 }
                       on ((Func<int, int>)(a => a))(a) 
                       equals b
                   select a + b;
```
Observed: No errors

Expected: An error about lambda parameter shadowing a range variable in scope
                       on ((Func<int, int>)(**a** => a))(a) 

This is going to affect pattern variable declarations and out variable declarations.

**Comment by AlekseyTs** [comment]
This is an issue with Linq implementation.

**Comment by AlekseyTs** [comment]
Note that this query is detecting the name conflict:
```
        var res2 = from a in new[] { 1 }
                   select ((Func<int, int>)(a => a))(a)
```

It looks like there is something wrong with conflict detection against transparent identifiers.

**Comment by gafter** [comment]
It isn't clear from the language specification (quoted below) that this situation is supposed to be an error. The required diagnostics apply to locals (and constants and parameters), but the spec does not include range variables (i.e. fields that are members of a transparent identifier) in the category of names subject to the conflict rule. Intuitively such an error would seem desirable, but it would require a language spec change. No such spec change is currently proposed for C# 7, so I'll assign this to a future release for tracking.

It is possible that we will elect not to make the language change, because as the original post illustrates, it would be a breaking change.

> - •	Each method declaration, indexer declaration, operator declaration, instance constructor declaration and anonymous function creates a new declaration space called a _**local variable declaration space**_. Names are introduced into this declaration space through formal parameters (_fixed-parameters_ and _parameter-arrays_) and _type-parameters_. The body of the function member or anonymous function, if any, is considered to be nested within the local variable declaration space. It is an error for a local variable declaration space and a nested local variable declaration space to contain elements with the same name. Thus, within a nested declaration space it is not possible to declare a local variable or constant with the same name as a local variable or constant in an enclosing declaration space. It is possible for two declaration spaces to contain elements with the same name as long as neither declaration space contains the other.
> - •	Each _block_ or _switch-block_ , as well as a _for_, _foreach_ and _using_ statement, creates a local variable declaration space for local variables and local constants. Names are introduced into this declaration space through _local-variable-declarations_ and _local-constant-declarations_, and by declarations in the headers of _foreach-statements_ and _specific-catch-clauses_. Note that blocks that occur as or within the body of a function member or anonymous function are nested within the local variable declaration space declared by those functions for their parameters. Thus it is an error to have e.g. a method with a local variable and a parameter of the same name.

**Comment by gafter** [comment]
@MadsTorgersen This issue demonstrates that members of a transparent identifier are not currently subject to the single-definition rule (in the spec and implementation). Should they be?

**Comment by AlekseyTs** [comment]
Here is an example with a local declaration rather than a lambda parameter:
```
        var res4 = from a in new[] { 1 }
                   from x2 in new[] { 0 }
                   join b in new[] { 1 }
                       on ((Func<int, int>)((c) =>
                       {
                           var a = c;
                           return a;
                       }))(a)
                       equals b
                   select a + b;
```

**Comment by gafter** [comment]
This example also falls afoul of the hole in the current language specification that members of transparent identifiers are not subject to the single definition rule.

**Comment by gafter** [comment]
I am closing this as by design, since the current language specification does not require an error here, and there is no proposal or champion for a proposal to add anything to the specification to require such an error.
